### DISCUSSION TITLE
Set up Immich as https web server

### DISCUSSION THREAD

**Discussion by BenjaminCz0101** [immich-app/immich#6894] [Q&A] [answered]
Hello
I have it running under proxmox as an lxc container and would like to run it as an nginx web server (https locally with a self-signed certificate). Is there a guide to this somewhere? I haven't found anything useful.

**Comment by danieldietzler** [comment, marked as answer]
Maybe [this](https://immich.app/docs/guides/remote-access#option-3-reverse-proxy) helps. Specifically, there is also an example nginx config referenced

**Comment by BenjaminCz0101** [comment]
Special thanks !

**Comment by BenjaminCz0101** [comment]
Hi
works for me with the config:

`server {

    listen 80;

    server_name immich;
	
	# https://github.com/immich-app/immich/blob/main/nginx/templates/default.conf.template#L28
    client_max_body_size 50000M;

    root /usr/share/nginx;

    access_log off;

    location /.well-known/acme-challenge/ { allow all; }

    location / { return 301 https://$host$request_uri; }

}


server {

    listen 443 ssl http2;

    server_name immich;

    access_log off;


    ssl_certificate     /opt/cert/immich.home.crt;

    ssl_certificate_key /opt/cert/immich.home.key;


    root /usr/share/nginx;

    location / {

        add_header Cache-Control no-cache;


        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_set_header X-Real-IP $remote_addr;

        proxy_set_header Host $http_host;

        proxy_pass http://192.168.178.xx:2283;

    }

}`


But when i give the ip http://192.168.178.xxx:2283 in the browser cames immich with http. How can I configure it to prevent this so that I can only access https.

**Reply by mmomjian** [reply]
Try this guide: [link](https://serversforhackers.com/c/redirect-http-to-https-nginx)

**Comment by BenjaminCz0101** [comment]
Hi
I have configured it, but still when I call http://192.168.178.xxx:2283 I am not redirected to https. Don't I have to change something in docker-compose so that it only runs locally on the server?

When i try https://192.168.178.xxx it works for https but Show´s Server offline ...

![Screenshot 2024-02-04 143913](https://github.com/immich-app/immich/assets/122832214/6bedcf98-201c-4853-83b6-fa9f69875113)


Here is my config:

`version: "3.8"

name: immich

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: [ "start.sh", "immich" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    ports:
      - 2283:3001
    depends_on:
      - redis
    restart: always

  immich-microservices:
    container_name: immich_microservices
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    # extends: # uncomment this section for hardware acceleration - see https://immich.app/docs/features/hardware-transcoding
    #   file: hwaccel.transcoding.yml 
    #   service: cpu # set to one of [nvenc, quicksync, rkmpp, vaapi, vaapi-wsl] for accelerated transcoding
    command: [ "start.sh", "microservices" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    depends_on:
      - redis
    restart: always

  immich-machine-learning:
    container_name: immich_machine_learning
    # For hardware acceleration, add one of -[armnn, cuda, openvino] to the image tag.
    # Example tag: ${IMMICH_VERSION:-release}-cuda
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    # extends: # uncomment this section for hardware acceleration - see https://immich.app/docs/features/ml-hardware-acceleration
    #   file: hwaccel.ml.yml
    #   service: cpu # set to one of [armnn, cuda, openvino, openvino-wsl] for accelerated inference - use the `-wsl` version for WSL2 where applicable
    volumes:
      - model-cache:/cache
    env_file:
      - .env
    restart: always

  redis:
    container_name: immich_redis
    image: redis:6.2-alpine@sha256:afb290a0a0d0b2bd7537b62ebff1eb84d045c757c1c31ca2ca48c79536c0de82
    restart: always


volumes:
  model-cache:
`
and my Nginx Config:


server {

    listen 80;

    server_name immich;
	
	# https://github.com/immich-app/immich/blob/main/nginx/templates/default.conf.template#L28
    client_max_body_size 50000M;

    root /usr/share/nginx;

    access_log off;

    location /.well-known/acme-challenge/ { allow all; }

    location / { return 301 https://$host$request_uri; }

}


server {

    listen 443 ssl http2;

    server_name immich;

    client_max_body_size 50000M;

    access_log off;


    ssl_certificate     /opt/cert/immich.home.crt;

    ssl_certificate_key /opt/cert/immich.home.key;


    root /usr/share/nginx;

    location / {

        add_header Cache-Control no-cache;


        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_set_header X-Real-IP $remote_addr;

        proxy_set_header Host $http_host;

        proxy_pass http://192.168.178.1xx:2283;

    }

}

**Comment by BenjaminCz0101** [comment]
OK found the Problem it was:
server_name immich;

change to:

server_name _;

But when i try https immich shows me no Server Status. Is the Problem i have only the port 2283 and not 3001 set ?

![Screenshot 2024-02-04 143913](https://github.com/immich-app/immich/assets/122832214/6cb001bf-33e4-4d29-94fe-79f44a0dad90)

**Reply by bo0tzz** [reply]
Make sure your proxy allows websockets, that usually causes this issue.

**Comment by BenjaminCz0101** [comment]
Thanks!! 
This Wroks for me: 

But i only need the port 2283 in my proxy manager to immich works fine right ?

`
server {

    listen 80;

    server_name _;
	
	# https://github.com/immich-app/immich/blob/main/nginx/templates/default.conf.template#L28
    client_max_body_size 50000M;

    root /usr/share/nginx;

    access_log off;

    location /.well-known/acme-challenge/ { allow all; }

    location / { return 301 https://$host$request_uri; }

}


server {

    listen 443 http2 ssl;
    listen [::]:443 http2 ssl;

    server_name 192.168.178.xx;

    client_max_body_size 50000M;

    ssl_certificate /opt/cert/immich.home.crt;
    ssl_certificate_key /opt/cert/immich.home.key;
    #ssl_dhparam /etc/ssl/certs/dhparam.pem;


    ########################################################################
    # from https://cipherli.st/                                            #
    # and https://raymii.org/s/tutorials/Strong_SSL_Security_On_nginx.html #
    ########################################################################

   


    ##################################
    # END https://cipherli.st/ BLOCK #
    ##################################


    root /usr/share/nginx/html;

    location / {
      proxy_set_header        Host $host;
      proxy_set_header        X-Real-IP $remote_addr;
      proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header        X-Forwarded-Proto $scheme;




      # Fix the “It appears that your reverse proxy set up is broken" error.
      proxy_pass          http://192.168.178.xx:2283;
      proxy_read_timeout  90;




      # WebSocket support
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";


    }


    error_page 404 /404.html;
    location = /404.html {
    }


    error_page 500 502 503 504 /50x.html;
    location = /50x.html {

    }
}
`

![Screenshot 2024-02-04 210442](https://github.com/immich-app/immich/assets/122832214/355a98d8-9d18-46d6-9b17-f5651efa7847)

**Comment by BenjaminCz0101** [comment]
Hello,
So everything is working so far, but the redirection isn't working yet. So https is activated and everything works. Only if I enter the IP address in the browser, in this case 192.168.178.xx:2283, will I not be redirected. If I just enter 192.168.178.xx I will be redirected to https://192.168.178.xx. What am I doing wrong ?

**Comment by BenjaminCz0101** [comment]
Hello,
I've tried everything possible but can't figure out what I did wrong. I also have an lxc container running the Docker Immich containers.

I then installed Nginx on the lxc container (no Docker, no Docker container!) and have now activated https. Also works, but when I go to 192.168.178.xx:2283 I don't get redirected to https. Do I have to change the port from 2283:3001 to 80 in Docker-Compose.yml?

**Comment by cyl0012345** [comment]
@BenjaminCz0101 
> But when i give the ip http://192.168.178.xxx:2283/ in the browser cames immich with http. How can I configure it to prevent this so that I can only access https.

You can update the `docker-compose.yml`, change "2283:3001" to "127.0.0.1:2283:3001" so that the docker will no longer expose the port to the public. 
After change you can run `ps | grep docker` to verify.
